### ISSUE TITLE
OWA ontop of Postgresql

### ISSUE THREAD

**Issue by archevel** [padams/Open-Web-Analytics#149]
I am trying to setup owa ontop of postgres instead of mysql. After a bit of digging I realized that I need a `owa_db_postgres.php` file in `plugins/db` if I set `OWA_DB_TYPE` to `postgres`. I've done a bit of searching in the repo but not found anything. Any advice on getting owa running ontop of postgres would be appreciated.

**Comment by padams** [comment]
You'd need to write a postgresql plugin class. Which for the most part is simply just defining PostgreSQL sql templates. See the MySQL plugin as s guide.

**Comment by archevel** [comment]
I had a quick look at creating a postgres plugin. Writing a `owa_db_postgres` class guided by `owa_db_mysql` seem straight forward. However, the data types defined is harder to map. Here are the issues I've come across after a quick stab at the problem:

- The definition of tables with auto incremented fields is different, which I think will cause problems. Postgres uses `SERIAL` to define an auto incrementing integer field whereas mysql uses `AUTO_INCREMENT` in conjunction with a `INT`. 
- Indexes are also created separately from the tables in postgres, i.e. not simply b adding `KEY` to the column definition.
- Mysql's `BLOB` is slightly different to posgresql's `BYTEA`), which might (?) cause problems depending on how it is used. 

It is probably possible to work around these issues by overriding `owa_db->createTable($entity)`, but since that function the `owa_entity` and `owa_dbColumn` are also tied to mysql syntax to genereate the definitions it doesn't seem as straight forward as I initaly expected...

**Comment by padams** [comment]
sounds like it would require some additional refactoring.

**Comment by bernd-wechner** [comment]
Shame this didn't happen. Seems important to me in this day and age to mainatian a certain backend database agnosticism. I mean if I wanted to put OWA on a machine already running postgresql (just if, ha, ha) why would i wants a MySQL instance running beside it? Just saying. A real shame OWA isn't easier to plug into any SQL database.

**Comment by padams** [comment]
It's totally possible to do this. just needs someone to code the db plugin for postgresql. 

The generic ORM/DAL libraries that give you built in agnosticism are good for CRUD queries but not for complex joins and indexing that OWA requires. 

Which is why OWA has its own db abstraction layer that can be extended to support any specific db via a plugin.

**Comment by bernd-wechner** [comment]
But didn't  @archevel fail to achieve this concluding that changes to the code base outside of the plugin were needed?

> It is probably possible to work around these issues by overriding `owa_db->createTable($entity)`, but since that function the `owa_entity` and `owa_dbColumn` are also tied to mysql syntax to genereate the definitions it doesn't seem as straight forward as I initaly expected...

Just curious.  What are your thoughts on those conclusions? 

Paraphrased, is there a complete list of all the MYSQL syntax dependent code that needs generalisation? And if any of it lies outside of `owa_db_mysql.php` should they be moved there to begin with?

I see not mentions for example of `createTable`, `owa_entity` or `owa_dbColumn` in `owa_db_mysql.php`, all of which @archevel identified as containing MYSQL specific code.

**Comment by padams** [comment]
I'm not sure i follow. All the MySQL specific syntax string templates are in `owa_db_mysql.php`. 

The Entity and column classes just sprintf values into those templates. I haven't dug into how the postreql create table syntax works but I'm pretty sure you can specify the primary key and unique keys along with the table create statement, no? 

If so then I think the matter can be confined to changing the template string for `OWA_SQL_CREATE_TABLE`. If the syntax is wildly different I suppose maybe it could require a change to the underlaying sprintf call but I'd need someone to lay the syntax out side by side in order for me to really tell.

If you'd like to work on this, then I suggest  making a new enhancement ticket for it where we can put investigations/conversations.  thanks.

**Comment by bernd-wechner** [comment]
Thanks for the reflections. I'm a tad time poor, but I have just forked the repo and cloned it and I'll take a look. Could well be that @archevel overestimated the complexity of it. Not sure. On a quick look I can see that table_options may pose a slight issue. But hey, I've got it on hand now and may take a look later as time permits as I have a Postgresql database running here and can debug fairly well (an excuse to set up a PHP debugging environment perhaps). We shall see. It would certanily be useful to me if OWA supported postgresql and useful to OWA if it could plug into existing postresql infrastructure easily.

I'll set it aside for tinkering now and come back if I find time to do that. Am in midst fo a server migration and part of that took me back to the question of analytics and I've been unhappy with AWstats. So ...
